docs/decisions/0004-iframe-credit-card-handling.rst

```diff
@@ -0,0 +1,47 @@
+
+Cybersource also provides an integration method called "Silent Order Post" (SOP) that allows merchants to put a <form> element on the basket page that directly posts to Cybersource allowing the merchant to avoid sending the sensitive data to back-end servers. This has been the method that frontend-app-payment has been using since it was initially created.
+
+Cybersource also provides an SDK called the "Flex Microform SDK" that allows merchants (like Open edX sites) to easily use the iframe integration method to embed the sensitive widgets directly into existing basket pages in a way that is transparent to the user.
+
+Decision
+--------
+
+frontend-app-payment will now use the Flex Microform SDK to handle credit card payments instead of SOP.
+
+Consequences
+------------
+
+1. The lifecycle of the page changes since we need to load and initialize the SDK in order to render the credit card number and CVN code form fields.
+2. The form fields are no longer part of the payment form, instead they are iframe elements. We have to do some work to make them behave like normal form fields when they aren't. For example: we style them to look like normal fields, we have special validation error handling to mimic the other fields etc.
+3. The SDK requires us to generate a "capture context" in order to initialize the fields. That capture context is only valid for 15 minutes. We need to refresh it every 15 minutes. When we refresh, we have to clear out any data entered into the credit card number and CVN fields. Additionally, the capture context must be generated using sensitive secrets. These secrets must be managed on the back-end so frontend-app-payment must make an AJAX request to fetch the capture context prior to rendering the credit card number and CVN fields.
+4. If Cybersource has a bug in their SDK javascript, it can prevent users from checking out.
+5. When the user clicks "Place Order" their credit card information is transformed into a transient payment token (a JWT) that can be used in lieu of their credit card number to authorize payments. However, the SOP API does not accept this type of token, so we had to switch over to using the Cybersource REST API to authorize the payment. This means that instead of direct POSTing the data to Cybersource via an HTML form, we instead make an AJAX request to the ecommerce back-end which, in turn, makes a synchronous HTTP request to Cybersource's servers.
```

Comment:
Removing SOP means that we have to handle automatic refunds if Cybersource denies but the CC processor accepts.

Reply:
Maybe worth mentioning as well that the backend lifecycle changes, because it's no longer waiting for a callback from Cybersource. Instead, it's making a call to cybersource and then processing the response to that call. That's different than all the other current payment providers.